exec/cnex/nstring.c

```diff
@@ -208,6 +222,99 @@ size_t string_findChar(TString *self, char c)
     return NPOS;
 }
 
+int64_t string_findFirstOf(TString *self, size_t pos, const char *str)
+{
+    if (self->data == NULL || self->length == 0 || pos > self->length || str == NULL) {
+        return -1;
+    }
+
+    for (size_t l = pos; l < self->length; l++) {
+        if (strchr(str, self->data[l]) != NULL) {
+            return l;
+        }
+    }
+    return -1;
+}
+
+static inline BOOL isNotOneOf(TString *self, char c)
+{
+    for (size_t i = 0; i < self->length; i++) {
+        if (self->data[i] == c) {
+            return FALSE;
+        }
+    }
+    return TRUE;
+}
+
+int64_t string_findFirstNotOf(TString *self, size_t pos, TString *p)
+{
+    if (self->data == NULL || self->length == 0 || p == NULL || p->data == NULL || pos > self->length) {
+        return -1;
+    }
+
+    // If a string of no chars is passed, should we return 0, or -1?  Technically, we didn't find anything,
+    // so the first char we didn't find would be 0.
+    if (p->length == 0) {
+        return 0;
+    }
+
+    for (size_t l = pos; l < self->length; l++) {
+        if (isNotOneOf(p, self->data[l])) {
+            return l;
+        }
+    }
+    return -1;
+}
+
+int64_t string_findLastNotOf(TString *self, TString *p)
+{
+    if (self->data == NULL || self->length == 0) {
+        return -1;
+    }
+
+    if (p == NULL || p->data == NULL || p->length == 0) {
+        return self->length;
+    }
+
+    for (size_t l = self->length; l > 0; l--) {
+        if (isNotOneOf(p, self->data[l-1])) {
+            return l;
+        }
+    }
+    return -1;
+}
+
+int64_t string_findString(TString *self, size_t pos, TString *p)
+{
+    int64_t r = -1;
+    if (p == NULL || p->data == NULL || p->length == 0) {
+        return r;
+    }
+    if (self == NULL || self->data == NULL || self->length == 0) {
+        return r;
+    }
+    if (pos > self->length || p->length > self->length || p->length + pos > self->length) {
+        return r;
+    }
+
+    for (size_t l = pos; l < ((self->length - p->length) + 1); l++) {
+        if (self->data[l] == p->data[0]) {
+            r = l;
+            BOOL bFound = TRUE;
+            for (size_t n = 0; n < p->length; n++) {
```

Comment:
It's redundant to do this comparison for n = 0 because you already know the first character matches.

Reply:
Yea, I knew that.  I guess I was just doing lazy comparison, by re-comparing the first char.  Fixed.